lib/rouge/lexers/php.rb

```diff
@@ -153,18 +165,19 @@ def self.detect?(text)
           end
         end
 
-        rule %r/(\d+\.\d*|\d*\.\d+)(e[+-]?\d+)?/i, Num::Float
-        rule %r/\d+e[+-]?\d+/i, Num::Float
-        rule %r/0[0-7]+/, Num::Oct
-        rule %r/0x[a-f0-9]+/i, Num::Hex
-        rule %r/\d+/, Num::Integer
+        rule %r/(?:(?:(?:\d+(?:_\d+)*)|(?:(?:(?:\d+(?:_\d+)*)?\.(?:\d+(?:_\d+)*))|(?:(?:\d+(?:_\d+)*)\.(?:\d+(?:_\d+)*)?)))e[+-]?(?:\d+(?:_\d+)*))/i, Num::Float
```

Comment:
Is there a way to simplify this rule? It seems awfully complicated just to match floats.

Reply:
I follow PHP's lexer:
```
LNUM	[0-9]+(_[0-9]+)*
DNUM	({LNUM}?"."{LNUM})|({LNUM}"."{LNUM}?)
EXPONENT_DNUM	(({LNUM}|{DNUM})[eE][+-]?{LNUM})
```
This rule matches EXPONENT_DNUM.

Can I do the same? I mean:
```ruby
lnum = /\d+(?:_\d+)*/
dnum = /(?:#{lnum}?\.#{lnum})|(?:#{lnum}\.#{lnum}?)/
rule %r/#{dnum}|((#{lnum}|#{dnum})e[+-]?#{lnum})/i, Num::Float
```

Reply:
I'd suggest:

```suggestion
rule %r/(\d[_\d]*)?\.(\d[_\d]*)?(e[+-]?\d[_\d]*)?/i, Num::Float
rule %r/\d[_\d]*/, Num::Integer
```

That'll include some invalid numbers on the margins but syntactic correctness is a non-goal of Rouge so I think it's worth it for the simplicity.

How does that look to you?

Reply:
Ok, I'll do that way.

But we need to keep `rule %r/0x[a-f0-9]+/i, Num::Hex`.

Reply:
To be consistent, should I rewrite the following:
```ruby
        rule %r/0[0-7]+(?:_[0-7]+)*/, Num::Oct
        rule %r/0b[01]+(?:_[01]+)*/i, Num::Bin
        rule %r/0x[a-f0-9]+(?:_[a-f0-9]+)*/i, Num::Hex
```
as:
```ruby
        rule %r/0[0-7][0-7_]*/, Num::Oct
        rule %r/0b[01][01_]*/i, Num::Bin
        rule %r/0x[a-f0-9][a-f0-9_]*/i, Num::Hex
```

Reply:
Yeah, I think that's a good point. I've incorporated those changes.